src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -10,26 +10,30 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 {
     public class IOQueue : PipeScheduler, IThreadPoolWorkItem
     {
-        private readonly object _workSync = new object();
         private readonly ConcurrentQueue<Work> _workItems = new ConcurrentQueue<Work>();
-        private bool _doingWork;
+        private int _doingWork;
 
         public override void Schedule(Action<object> action, object state)
         {
-            var work = new Work
+            _workItems.Enqueue(new Work
             {
                 Callback = action,
                 State = state
-            };
+            });
 
-            _workItems.Enqueue(work);
+            // Ensure Enqueue happens prior to checking _doingWork.
+            // As they are two different memory locations, we insert a barrier to guarantee ordering.
+            Thread.MemoryBarrier();
```

Comment:
Why not just always attempt the exchange instead?

Reply:
Can change it to `CompareExchange` I suppose

Reply:
Can change it to `CompareExchange` I suppose as that will force a barrier